Files/ViewModels/ItemViewModel.cs

```diff
@@ -460,143 +475,193 @@ public void CancelExtendedPropertiesLoading()
             loadPropsCTS = new CancellationTokenSource();
         }
 
-        public void OrderFiles(IList<ListedItem> orderedList = null)
+        // apply changes immediately after manipulating on filesAndFolders completed
+        public async Task ApplyFilesAndFoldersChangesAsync()
         {
-            if (orderedList == null)
+            try
             {
-                orderedList = OrderFiles2(filesAndFolders);
+                await updateDataGridSemaphore.WaitAsync(addFilesCTS.Token);
             }
-            //_filesAndFolders.BeginBulkOperation();
-            for (var i = 0; i < orderedList.Count; i++)
+            catch (Exception ex) when (ex is OperationCanceledException || ex is ObjectDisposedException)
             {
-                if (i < filesAndFolders.Count)
+                return;
+            }
+
+            try
+            {
+                if (filesAndFolders == null || filesAndFolders.Count == 0)
                 {
-                    if (filesAndFolders[i] != orderedList[i])
+                    await CoreApplication.MainView.ExecuteOnUIThreadAsync(() =>
                     {
-                        filesAndFolders.Insert(i, orderedList[i]);
-                    }
+                        FilesAndFolders.Clear();
+                        IsFolderEmptyTextDisplayed = FilesAndFolders.Count == 0;
+                        UpdateDirectoryInfo();
+                    });
+                    return;
                 }
-                else
+
+                // ObservableCollection.CollectionChanged will cause UI update, which may cause
+                // significant performance degradation, so suppress ObservableCollection.CollectionChanged
+                // event here while loading items heavily
+                FilesAndFolders.BeginBulkOperation();
+                // After calling BeginBulkOperation, ObservableCollection.CollectionChanged is suppressed
+                // so modifies to FilesAndFolders won't trigger UI updates, hence below operations can be
+                // run safely without needs of dispatching to UI thread
+                await Task.Run(() =>
                 {
-                    filesAndFolders.Add(orderedList[i]);
-                }
+                    for (var i = 0; i < filesAndFolders.Count; i++)
+                    {
+                        if (addFilesCTS.IsCancellationRequested) return;
+                        if (i < FilesAndFolders.Count)
+                        {
+                            if (FilesAndFolders[i] != filesAndFolders[i])
```

Comment:
Here we should use value equality, reference won't match with list from cache

Reply:
I think reference equality is needed here, because the only thing `ApplyFilesAndFoldersChangesAsync` needs to do is to reflect changes from `filesAndFolders`, and all entries in `FilesAndFolders` are referenced from `filesAndFolders`.

Reply:
Yes, but what about changes in 'filesAndFolders'? I mean - when some item is loaded from cache and then the same file is loaded from filesystem their reference won't match. It would be ideal for the 'FilesAndFolder' not to refresh lost in that case

Reply:
As for now the case "some item is loaded from cache and then the same file is loaded from filesystem" doesn't exists. Change this may cause unexpected behavior in `LoadExtendedItemProperties` for it uses item from `FilesAndFolders` directly.

Reply:
I'm pretty sure at some point you have to check if cache entries are up to date. In that place value equality could potentially be better. 

I don't have access to bigger screen right now, will take a look at this tomorrow

Reply:
I went through the code and here's what i got:
- First - you are loading files from cache (if any)
- Second - you are running enumeration though a filesystem to load files from drive
At any point you do not clean `filesAndFolders` list, so when files are loaded from drive at some point they are compared to the files already loaded from cache (in function `ApplyFilesAndFoldersChangesAsync`. Since files loaded from cache are different reference then those loaded from filesystem - it will always call `FilesAndFolders.Insert(i, filesAndFolders[i]);` forcing an unnecessary refresh of this particular item (i think, again i'm not an UWP guy, so here's where i can be dead wrong). And i think this can (and should probably?) be avoided by using value equality - do not replace item read from cache if it's the same item (i think value equality can be based on path + type). @gave92 @yaichenbaum do you think that's reasonable?

Reply:
In 06b47f4 I just skipped entries which already loaded from cache, so I think it's not a issue anymore. Also, comparing by reference is quicker than comparing by value.

Reply:
Ok, i think that's fine. 

Comparing by reference is basically a comparison on int/long so it is faster, but i'm pretty sure we're talking about nanoseconds here ;)